### PR TITLE
[BEAM-10925] Add interfaces for SQL Java scalar UDF.

### PR DESCRIPTION
**Pull request by ibzib**
R: @amaliujia @kennknowles 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Review by amaliujia — CHANGES_REQUESTED** [review]
Mark interfaces as experimental?

**Comment by ibzib** [comment]
> Mark interfaces as experimental?

Good thing I broke up the PR, I totally missed that. Done.

**Review by amaliujia — APPROVED** [review]

**Comment by kennknowles** [comment]
Checking here: the interface for a scalar UDF is "has a method annotated `@ApplyMethod`".

The downside for an interface like this is that it cannot be instantiated dynamically. I cannot take some existing object or code and programmatically turn it into a scalar function. We have hit this problem with people who want to programmatically create DoFn but are unable to do so. The Euphoria DSL for example and also you cannot take a DoFn and wrap it into a related DoFn. Annotation-driven things don't compose well.

It also means that if you have a handle on a `ScalarFn` you can only invoke it via reflection. If I recall, your full change includes the reflection helpers to make that invocation easy, the same way we map DoFn to (DoFnInvoker, DoFnSignature). But I want to say that it was probably a mistake to treat DoFnInvoker as an implementation detail and DoFn as the main object. Instead, I would now view (DoFnInvoker, DoFnSignature) pair as the fundamental composable object, and annotated DoFn just a user interface for defining one. It is essentially an embedded programming language: DoFn is the syntax, DoFnSignature is the type, DoFnInvoker is the semantic object.

I'm not sure I was clear before, but I really would recommend taking this approach for SQL. Perhaps this does not make sense because actually the main thing you do with these is pass them to Calcite to plug into linq4j?

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/ApplyMethod.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import java.lang.annotation.Target;
+import org.apache.beam.sdk.annotations.Experimental;
+
+/**
+ * Annotates the single method in {@link ScalarFn} that is to be applied to SQL function arguments.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+@Experimental
+public @interface ApplyMethod {}
```
Comment (apilloud) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/ApplyMethod.java:32]: Should this interface be nested within ScalarFn?

**Comment by apilloud** [comment]
This should be moved into it's own package without any dependencies on Beam. This will enable users to write UDFs without taking a dependency on Beam.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/ApplyMethod.java (hunk 1 above)
Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/ApplyMethod.java:32]: Yes (if we do decide to go with the annotations route).

Comment (ibzib) [reply sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/ApplyMethod.java]: Done.

**Comment by ibzib** [comment]
> This should be moved into it's own package without any dependencies on Beam. This will enable users to write UDFs without taking a dependency on Beam.

Done.

**Comment by ibzib** [comment]
> The downside for an interface like this is that it cannot be instantiated dynamically. I cannot take some existing object or code and programmatically turn it into a scalar function. We have hit this problem with people who want to programmatically create DoFn but are unable to do so. The Euphoria DSL for example and also you cannot take a DoFn and wrap it into a related DoFn. Annotation-driven things don't compose well.
> 
> It also means that if you have a handle on a `ScalarFn` you can only invoke it via reflection. If I recall, your full change includes the reflection helpers to make that invocation easy, the same way we map DoFn to (DoFnInvoker, DoFnSignature). But I want to say that it was probably a mistake to treat DoFnInvoker as an implementation detail and DoFn as the main object. Instead, I would now view (DoFnInvoker, DoFnSignature) pair as the fundamental composable object, and annotated DoFn just a user interface for defining one. It is essentially an embedded programming language: DoFn is the syntax, DoFnSignature is the type, DoFnInvoker is the semantic object.
> 
> I'm not sure I was clear before, but I really would recommend taking this approach for SQL. Perhaps this does not make sense because actually the main thing you do with these is pass them to Calcite to plug into linq4j?

I think it's important to be aware of these drawbacks, but I don't think they necessarily apply here. The existing DoFn implementation works well for the majority of regular users, and I'm not anticipating the kind of advanced usage we have with DoFns. SQL UDFs are a much more limited use cases than DoFn, and DSLs are best implemented at a lower abstraction layer like DoFn.

It'd be nice to have lambdas, but I think this marker interface is the best suited to the problem (because UDFs have indefinite arity and typing) for the majority of, if not all potential users.

**Comment by ibzib** [comment]
Run Java PreCommit

**Comment by ibzib** [comment]
Run Python_PVR_Flink PreCommit

**Review by amaliujia — APPROVED** [review]
Thanks for moving this interface to its own package. This interface is experimental so we still be able to refine it later depending on requests or requirements.

File: sdks/java/extensions/sql/udf/src/main/java/org/apache/beam/sdk/extensions/sql/udf/ScalarFn.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * </code></pre>
+ */
+public abstract class ScalarFn implements Serializable {
+  /**
+   * Annotates the single method in a {@link ScalarFn} implementation that is to be applied to SQL
+   * function arguments.
+   */
+  @Retention(RetentionPolicy.RUNTIME)
+  @Target(ElementType.METHOD)
+  public static @interface ApplyMethod {}
```
Comment (amaliujia) [inline sdks/java/extensions/sql/udf/src/main/java/org/apache/beam/sdk/extensions/sql/udf/ScalarFn.java:49]: SHould be merged as @Experimental?

File: sdks/java/extensions/sql/udf/build.gradle [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+plugins { id 'org.apache.beam.module' }
+
+applyJavaNature(automaticModuleName: 'org.apache.beam.sdk.extensions.sql.udf')
```
Comment (apilloud) [inline sdks/java/extensions/sql/udf/build.gradle:21]: Might be worth adding a comment about not adding dependencies here.

**Review by apilloud — APPROVED** [review]
LGTM. Thanks!

File: sdks/java/extensions/sql/udf/build.gradle (hunk 3 above)
Comment (ibzib) [reply sdks/java/extensions/sql/udf/build.gradle:21]: Done.

File: sdks/java/extensions/sql/udf/src/main/java/org/apache/beam/sdk/extensions/sql/udf/ScalarFn.java (hunk 2 above)
Comment (ibzib) [reply sdks/java/extensions/sql/udf/src/main/java/org/apache/beam/sdk/extensions/sql/udf/ScalarFn.java:52]: Experimental itself is a part of Beam java core: https://github.com/apache/beam/blob/486d23007dac513f9f90bb08c2a039cd72a7ec94/sdks/java/core/src/main/java/org/apache/beam/sdk/annotations/Experimental.java

I added a note to the javadoc instead.

**Comment by ibzib** [comment]
Run Python PreCommit

**Pull request merged**